system-addon/lib/DownloadsManager.jsm

```diff
@@ -0,0 +1,146 @@
+      return null;
+    }
+    let items = this.download.target.path.split(".");
+    return items[items.length - 1].toUpperCase();
+  }
+
+  downloadsCmd_copyLocation() {
+    let clipboard = Cc["@mozilla.org/widget/clipboardhelper;1"].getService(Ci.nsIClipboardHelper);
+    clipboard.copyString(this.download.source.url);
+  }
+}
+
+this.DownloadsManager = class DownloadsManager {
+  constructor(store) {
+    this._downloadData = null;
+    this._window = null;
+    this._store = null;
+    this._viewableDownloadItems = new Map();
+  }
+
+  formatDownload(element) {
+    const downloadedItem = element.download;
+    let description;
+    if (element.fileType) {
+      // If we have a file type: '1.5 MB — PNG'
+      description = `${element.sizeStrings.stateLabel} \u2014 ${element.fileType}`;
+    } else {
+      // If we do not have a file type: '1.5 MB'
+      description = `${element.sizeStrings.stateLabel}`;
+    }
+    return {
+      hostname: new URL(downloadedItem.source.url).hostname,
+      url: downloadedItem.source.url,
+      path: downloadedItem.target.path,
+      title: element.displayName,
+      description,
+      referrer: downloadedItem.source.referrer,
+      date_added: downloadedItem.endTime
+    };
+  }
+
+  init(store) {
+    this._store = store;
+    this._browser = Services.appShell.hiddenDOMWindow;
+    this._downloadData = DownloadsCommon.getData(this._browser.ownerGlobal, true, false, true);
+    this._downloadData.addView(this);
+  }
+
+  onDownloadAdded(download) {
+    const elem = new DownloadElement(download, this._browser);
+    const downloadedItem = elem.download;
+    if (!this._viewableDownloadItems.has(downloadedItem.source.url)) {
+      this._viewableDownloadItems.set(downloadedItem.source.url, elem);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  onDownloadRemoved(download) {
+    if (this._viewableDownloadItems.has(download.source.url)) {
+      this._viewableDownloadItems.delete(download.source.url);
+      this._store.dispatch({type: at.DOWNLOAD_CHANGED});
+    }
+  }
+
+  get downloads() {
+    let results = [];
+    for (const url of this._viewableDownloadItems.keys()) {
+      const elem = this._viewableDownloadItems.get(url);
+      // We only want downloads that are less than 36 hours old
+      if (elem && (elem.downloadedSince < RECENT_DOWNLOAD_THRESHOLD)) {
+        const formattedDownloadForHighlights = this.formatDownload(elem);
+        results.push(formattedDownloadForHighlights);
+      }
+    }
+    return results.sort((a, b) => a.date_added < b.date_added);
+  }
+
+  uninit() {
+    if (this._downloadData) {
+      this._downloadData.removeView(this);
+      this._downloadData = null;
+    }
+  }
+
+  onAction(action) {
+    let downloadsCmd;
+    switch (action.type) {
+      case at.COPY_DOWNLOAD_LINK:
+        downloadsCmd = "copyLocation";
+        break;
+      case at.GO_TO_DOWNLOAD_PAGE:
+        downloadsCmd = "openReferrer";
```

Comment:
This seems to trigger an exception `this.element.ownerGlobal.openURL is not a function DownloadsViewUI.jsm:453`﻿ 
https://searchfox.org/mozilla-central/rev/59a9a86553e9bfd9277202748ff791fd9bc0713b/browser/components/downloads/DownloadsViewUI.jsm#452-453